src/cpu/instr.rs

```diff
@@ -40,30 +74,258 @@ struct Operation {
+        });
+        m.insert(Opcode::DIV as u8, Operation {
+           execute: |enc, cpu| {
+               cpu.state.reg[enc.rd as usize] = cpu.state.reg[enc.rs1 as usize] / cpu.state.reg[enc.rs2 as usize];
+               cpu.state.pc = cpu.state.pc+1;
+           },
+           repr: |enc| format!("div r{0}, r{1}, r{2}", enc.rd, enc.rs1, enc.rs2),
+        });
+        m.insert(Opcode::MUL as u8, Operation {
+            execute: |enc, cpu| {
+                cpu.state.reg[enc.rd as usize] = cpu.state.reg[enc.rs1 as usize] * cpu.state.reg[enc.rs2 as usize];
+                cpu.state.pc = cpu.state.pc+1;
+            },
+            repr: |enc| format!("mul r{0}, r{1}, r{2}", enc.rd, enc.rs1, enc.rs2),
+        });
+        m.insert(Opcode::MOD as u8, Operation {
+            execute: |enc, cpu| {
+                cpu.state.reg[enc.rd as usize] = cpu.state.reg[enc.rs1 as usize] % cpu.state.reg[enc.rs2 as usize];
+                cpu.state.pc = cpu.state.pc+1;
+            },
+            repr: |enc| format!("mod r{0}, r{1}, r{2}", enc.rd, enc.rs1, enc.rs2),
+        });
+        m.insert(Opcode::CMP as u8, Operation {
+            execute: |enc, cpu| {
+                let _out: u32 = cpu.state.reg[enc.rs1 as usize] as u32 - cpu.state.reg[enc.rs2 as usize] as u32;
+                cpu.state.flags = gen_flag(true, cpu.state.reg[enc.rs1 as usize] as u32, cpu.state.reg[enc.rs1 as usize] as u32, _out)
+            },
+            repr: |enc| format!("cmp r{}, r{}", enc.rs1, enc.rs2),
+        });
+        m.insert(Opcode::CMI as u8, Operation {
+            execute: |enc, cpu| {
+                let _out: u32 = cpu.state.reg[enc.rs1 as usize] as u32 - enc.imm as u32;
+                cpu.state.flags = gen_flag(true, cpu.state.reg[enc.rs1 as usize] as u32, enc.imm as u32, _out);
+            },
+            repr: |enc| format!("cmp r{}, {}", enc.rs1, enc.imm),
+        });
+        m.insert(Opcode::JAL as u8, Operation {
+            execute: |enc, cpu| {
+                cpu.state.reg[enc.rd as usize] = cpu.state.pc;
+                cpu.state.pc = enc.imm; 
+            },
+            repr: |enc| format!("jal {} {}",enc.rd, enc.imm),
+        });
+        m.insert(Opcode::JMP as u8, Operation {
+            execute: |enc, cpu| {
+                let jmp_code: u8 = ((enc.rs1 << 3) as u8) +  enc.rd as u8;
+                let cpu_flags: Flags = Flags::from_bits_truncate(cpu.state.flags);
+                let mut jump_condition_met: bool = false;
+
+                match jmp_code {
+                    0x0 => cpu.state.pc = enc.imm,
+                    0x1 => if cpu_flags.contains(Flags::C) { jump_condition_met = true },
+                    0x2 => if cpu_flags.contains(Flags::Z) { jump_condition_met = true },
+                    0x3 => if cpu_flags.contains(Flags::N) { jump_condition_met = true },
+                    0x4 => if !(cpu_flags.contains(Flags::N | Flags::Z)) { jump_condition_met = true },
```

Comment:
This would not work. (previous implemetation probably would not too). `.contains` check if all of the flags are set. Your condition would be false only if two flags are set at once, but in `~(N|Z)` expression only one of them needs to be true for the expression to be false. Replace with `!(contains(N) | contains(Z))`